File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    int canJump(vector<int>& nums) {
       vector<int> solution(nums.size());
       int runningCounter=0;
        if(nums.size()==1){     
            return true;
        }
       solution[nums.size()-1]= 0;
       runningCounter = 2; 
       if(nums.size()==2){
           if(nums[0]>=1){
               return true;
           }
            return false;
       }
       if(nums[nums.size()-2]==0){
        solution[nums.size()-2]=0;
       }
       else{
        solution[nums.size()-2]=1;
       }
       for(int it = nums.size()-3;it>=0;it--){
            if(nums[it]==0){
                solution[it]=0;
            }
            if(nums[it]>=runningCounter){
                solution[it]= 1;
            }
            else{
               int min = 50000;
               for(int j = it+1;j<=it+nums[it];j++){
                if(solution[j]>0 and solution[j]<min){
                    min=solution[j];
                }
               }
               solution[it]=min==50000?0:min+1;
            }
           runningCounter++;
       }
        
        return solution[0]>0?true:false;
    }
};

File: 0055-jump-game/0055-jump-game.py
class Solution:
    
    def rec(self, nums, availMoves, currentIndex):
        if self.dp.get(currentIndex) is not None:
            return self.dp[currentIndex]
        if currentIndex >= len(nums)-1:
            return True
        maxMovesAvailables = max(availMoves, nums[currentIndex])
        if maxMovesAvailables == 0:
            return False
        self.dp[currentIndex+1] = self.rec(nums, maxMovesAvailables - 1, currentIndex+1)
        self.dp[currentIndex+maxMovesAvailables] =  self.rec(nums, 0, currentIndex+maxMovesAvailables)
        return self.dp[currentIndex+1] or self.dp[currentIndex+maxMovesAvailables]
    
    def canJump(self, nums: List[int]) -> bool:
        self.dp = {}
        if len(nums)==1:
            return True
        self.rec(nums, 0, 0)
        if self.dp.get(len(nums)-1) is None:
            return False
        return self.dp.get(len(nums)-1)
        #Choice should be to whether jump the maximum or move one ahead and check the result